# Find the Factorial of a non-negative integer

In [1]:
def factorial(n):
    if n==1 or n==0:
        return 1
    else:
        return n * factorial(n-1)
n = 5
print(factorial(n))

120


* Input: n = 5 (Calculate 5!, or 5 factorial).
* First call: factorial(5) = 5 * factorial(4).
* Second call: factorial(4) = 4 * factorial(3).
* Third call: factorial(3) = 3 * factorial(2).
* Fourth call: factorial(2) = 2 * factorial(1).
* Base case: factorial(1) = 1.

factorial(2) = 2 * 1 = 2

factorial(3) = 3 * 2 = 6

factorial(4) = 4 * 6 = 24

factorial(5) = 5 * 24 = 120.


# Find the Exponential

In [2]:
def exponential(a, n):
    if n == 0:
        return 1
    else: 
        return a * exponential(a, n-1)
n = 5
a = 2
print(exponential(a, n))

32


* Input: a = 2, n = 5 (Calculate 2^5).
* First call: exponential(2, 5) = 2 * exponential(2, 4)
* Second call: exponential(2, 4) = 2 * exponential(2, 3).
* Third call: exponential(2, 3) = 2 * exponential(2, 2).
* Fourth call: exponential(2, 2) = 2 * exponential(2, 1).
* Fifth call: exponential(2, 1) = 2 * exponential(2, 0).
* Base case: exponential(2, 0) = 1.

exponential(2, 1) = 2 * 1 = 2

exponential(2, 2) = 2 * 2 = 4

exponential(2, 3) = 2 * 4 = 8

exponential(2, 4) = 2 * 8 = 16

exponential(2, 5) = 2 * 16 = 32.

# Sum of n numbers

In [3]:
def nsum(n):
    if n == 1: 
        return 1
    else: 
        return n+nsum(n-1)
n = 4
print(nsum(n))

10


    Input: n = 4 (Calculate the sum of the first 4 natural numbers: 1+2+3+41+2+3+4).

Step-by-Step Recursive Breakdown:

* First Call:
      nsum(4) = 4 + nsum(3)
* Second Call:
      nsum(3) = 3 + nsum(2)
* Third Call:
      nsum(2) = 2 + nsum(1)
* Base Case:
      nsum(1) = 1

    nsum(2) = 2 + 1 = 3
    nsum(3) = 3 + 3 = 6
    nsum(4) = 4 + 6 = 10

# Sum of a list of numbers

In [4]:
def list_sum(numslist):
    if len(numslist)==1:
        return numslist[0]
    else:
        return numslist[0] + list_sum(numslist[1:])
numslist = [2,4,5,6,7]
print(list_sum(numslist))

24


# Fibonacci sequence

In [5]:
def fibonacci(n):
    if n == 0:
        return 0
    elif n==1:
        return 1
    else:
        return fibonacci(n-1)+ fibonacci(n-2)
n = 6
print(fibonacci(n))

8


    Input: n = 6 (Calculate the 6th Fibonacci number).
    The Fibonacci sequence is: 0,1,1,2,3,5,8,…0,1,1,2,3,5,8,…
    Goal: Compute F(6)=F(5)+F(4)F(6)=F(5)+F(4) using recursion.
* First Call: fibonacci(6) = fibonacci(5) + fibonacci(4)
* Second Call: fibonacci(5) = fibonacci(4) + fibonacci(3)
* Third Call: fibonacci(4) = fibonacci(3) + fibonacci(2)
* Base Cases:

    fibonacci(1) = 1, fibonacci(0) = 0

    fibonacci(2) = fibonacci(1) + fibonacci(0) = 1 + 0 = 1
    fibonacci(3) = fibonacci(2) + fibonacci(1) = 1 + 1 = 2
    fibonacci(4) = fibonacci(3) + fibonacci(2) = 2 + 1 = 3
    fibonacci(5) = fibonacci(4) + fibonacci(3) = 3 + 2 = 5
    fibonacci(6) = fibonacci(5) + fibonacci(4) = 5 + 3 = 8

# Sum of Array Elements

In [6]:
def linearsum(data,n):
    if n == 0:
        return 0
    else:
        return data[n-1]+ linearsum(data,n-1)
print(linearsum([1,2,3,4,5,6,7,8],8))

36


    Input:
        data = [1, 2, 3, 4, 5, 6, 7, 8]
        n = 8 (Sum all elements in the list).

    Step-by-Step Recursive Breakdown:

* First Call:
    linearsum(data, 8) = data[7] + linearsum(data, 7)
    Add data[7] = 8 to the result of the next recursive call.

* Second Call:
    linearsum(data, 7) = data[6] + linearsum(data, 6)
    Add data[6] = 7 to the result of the next recursive call.

* Third Call:
    linearsum(data, 6) = data[5] + linearsum(data, 5)
    Add data[5] = 6 to the result of the next recursive call.

* Repeat Until Base Case:
    Base Case: linearsum(data, 0) = 0

    linearsum(data, 1) = data[0] + 0 = 1
    linearsum(data, 2) = data[1] + 1 = 2 + 1 = 3
    linearsum(data, 3) = data[2] + 3 = 3 + 3 = 6
    linearsum(data, 4) = data[3] + 6 = 4 + 6 = 10
    linearsum(data, 5) = data[4] + 10 = 5 + 10 = 15
    linearsum(data, 6) = data[5] + 15 = 6 + 15 = 21
    linearsum(data, 7) = data[6] + 21 = 7 + 21 = 28
    linearsum(data, 8) = data[7] + 28 = 8 + 28 = 36

# Linear Search in an array

# Binary Search

In [7]:
def binarySearch(data,low,high,key):
    if low>high:
        return -1
    else:
        mid =(low+high)//2
        if key == data[mid]:
            return mid
        elif key < data[mid]:
            return binarySearch(data,low,mid-1,key)
        else:
            return binarySearch(data,mid+1,high,key)
print(binarySearch([2,3,4,5,6,7,12,14,17,19,22,25,27, 28,33,37],0,15,22))

10


* Input:
        data = [2, 3, 4, 5, 6, 7, 12, 14, 17, 19, 22, 25, 27, 28, 33, 37]
        low = 0, high = 15, key = 22

* First Call:
  
  low = 0, high = 15 → mid = (0 + 15) // 2 = 7
  data[7] = 14. Since key > data[7], search the right half → Recursive call with low = 8, high = 15.

* Second Call:

  low = 8, high = 15 → mid = (8 + 15) // 2 = 11
  data[11] = 25. Since key < data[11], search the left half → Recursive call with low = 8, high = 10.

* Third Call:
  
  low = 8, high = 10 → mid = (8 + 10) // 2 = 9
  data[9] = 19. Since key > data[9], search the right half → Recursive call with low = 10, high = 10.

* Fourth Call:
  
  low = 10, high = 10 → mid = (10 + 10) // 2 = 10
  data[10] = 22. Since key == data[10], return 10.

# Converty integer to string in any base

In [8]:
def convert_to_string(n, base):
    convert_string = "0123456789ABCDEF"

    if n<base:
        return convert_string[n]
    else:
        return convert_to_string(n//base, base) + convert_string[n%base]
n = 2835
base = 16
print(convert_to_string(n,base))

B13


* Input: n = 2835, base = 16 (Convert 2835 to hexadecimal).
* First call: 2835 // 16 = 177, 2835 % 16 = 3 → Recursive call with n = 177.
* Second call: 177 // 16 = 11, 177 % 16 = 1 → Recursive call with n = 11.
* Third call (Base Case): n = 11 < 16 → Return convert_string[11] = 'B'.
* Combine results: 'B' + '1' + '3' = 'B13'. Output is 'B13'.



In [9]:
# Define a function named recursive_list_sum that calculates the sum of elements in a nested list
def recursive_list_sum(data_list):
    # Initialize a variable 'total' to store the cumulative sum
    total = 0
    
    # Iterate through each element in the input list
    for element in data_list:
        # Check if the current element is a list (nested list)
        if type(element) == type([]):
            # If the element is a list, recursively call the recursive_list_sum function on the element
            total = total + recursive_list_sum(element)
        else:
            # If the element is not a list, add its value to the total
            total = total + element

    # Return the total sum
    return total

# Print the result of calling the recursive_list_sum function with the input list [1, 2, [3,4], [5,6]]
print(recursive_list_sum([1, 2, [3, 4], [5, 6]]))

21


# Find the GCD of two integers

In [10]:
# Define a function named Recurgcd that calculates the greatest common divisor (GCD) of two numbers 'a' and 'b' using recursion and the Euclidean algorithm

def Recurgcd(a, b):
    # Determine the lower and higher values between 'a' and 'b'
    low = min(a, b)
    high = max(a, b)

    # Check if the lower value is 0 (base case for GCD calculation)
    if low == 0:
        # If the lower value is 0, return the higher value (GCD is the non-zero value)
        return high
    # Check if the lower value is 1 (base case for GCD calculation)
    elif low == 1:
        # If the lower value is 1, return 1 (GCD of any number with 1 is 1)
        return 1
    else:
        # If neither base case is met, recursively call the Recurgcd function
        # with the lower value and the remainder of the higher value divided by the lower value
        return Recurgcd(low, high % low)

# Print the result of calling the Recurgcd function with the input values 12 and 14
print(Recurgcd(12, 14))


2


# The value of 'a' to the power of 'b'

In [11]:
# Define a function named power that calculates the result of 'a' raised to the power of 'b'
def power(a, b):
    # Check if 'b' is 0 (base case for power function)
    if b == 0:
        # If 'b' is 0, return 1 (any number raised to the power of 0 is 1)
        return 1
    # Check if 'a' is 0 (base case for power function)
    elif a == 0:
        # If 'a' is 0, return 0 (0 raised to any power is 0)
        return 0
    # Check if 'b' is 1 (base case for power function)
    elif b == 1:
        # If 'b' is 1, return 'a' (any number raised to the power of 1 is the number itself)
        return a
    else:
        # If none of the base cases is met, recursively call the power function
        # to calculate 'a' multiplied by the result of 'a' raised to the power of 'b-1'
        return a * power(a, b - 1)

# Print the result of calling the power function with the input values 3 and 4
print(power(3, 4))


81


# Find the geometric sum up to 'n' terms

In [12]:
# Define a function named geometric_sum that calculates the geometric sum up to 'n' terms
def geometric_sum(n):
    # Check if 'n' equals 0, which is the base case for the geometric sum
    if n == 0:  # Corrected base case condition
        # If 'n' equals 0, return 1 as the geometric sum in this case is 1
        return 1
    else:
        # If 'n' is not 0, calculate the term in the geometric series (1 / 2^n) and add it to
        # the result of recursively calling the geometric_sum function with 'n - 1'
        return 1 / (pow(2, n)) + geometric_sum(n - 1)
# Print the result of calling the geometric_sum function with the input value 7
print(geometric_sum(7))
# Print the result of calling the geometric_sum function with the input value 4
print(geometric_sum(4))

1.9921875
1.9375


# Sum of a harmonic series upto n terms

In [13]:
# Define a function named harmonic_sum that calculates the harmonic sum up to 'n' terms
def harmonic_sum(n):
    # Check if 'n' is less than 2 (base case for the harmonic sum)
    if n < 2:
        # If 'n' is less than 2, return 1 (base case value for the harmonic sum)
        return 1
    else:
        # If 'n' is greater than or equal to 2, calculate the reciprocal of 'n'
        # and add it to the result of recursively calling the harmonic_sum function with 'n - 1'
        return 1 / n + harmonic_sum(n - 1)

# Print the result of calling the harmonic_sum function with the input value 7
print(harmonic_sum(7))

# Print the result of calling the harmonic_sum function with the input value 4
print(harmonic_sum(4))

2.5928571428571425
2.083333333333333


# The sum of the positive integers of n+(n-2)+(n-4)... (until n-x =< 0)

In [14]:
# Define a function named sum_series that calculates the sum of a series of numbers
def sum_series(n):
    # Check if 'n' is less than 1 (base case for the series)
    if n < 1:
        # If 'n' is less than 1, return 0 (base case value for the sum)
        return 0
    else:
        # If 'n' is greater than or equal to 1, calculate the sum of 'n' and
        # recursively call the sum_series function with 'n - 2'
        return n + sum_series(n - 2)

# Print the result of calling the sum_series function with the input value 6
print(sum_series(6))

# Print the result of calling the sum_series function with the input value 10
print(sum_series(10))

12
30
